core/web/auth/auth.go

```diff
@@ -1,135 +1,179 @@
-}
+var _ authMethod = AuthenticateBySession
 
 // AuthenticateByToken authenticates a User by their API token.
-func AuthenticateByToken(store AuthStorer, c *gin.Context) error {
+//
+// Implements authMethod
+func AuthenticateByToken(c *gin.Context, authr Authenticator) error {
 	token := &auth.Token{
 		AccessKey: c.GetHeader(APIKey),
 		Secret:    c.GetHeader(APISecret),
 	}
 
-	user, err := store.FindUser()
-	if errors.Is(err, sql.ErrNoRows) {
-		return auth.ErrorAuthFailed
-	} else if err != nil {
+	user, err := authr.FindUser()
+	if err != nil {
+		if errors.Is(err, sql.ErrNoRows) {
+			return auth.ErrorAuthFailed
+		}
+
 		return err
 	}
 
 	ok, err := clsessions.AuthenticateUserByToken(token, &user)
 	if err != nil {
 		return err
-	} else if !ok {
+	}
+	if !ok {
 		return auth.ErrorAuthFailed
 	}
+
 	c.Set(SessionUserKey, &user)
+
 	return nil
 }
 
-var _ authType = AuthenticateByToken
+var _ authMethod = AuthenticateByToken
 
-func AuthenticateBySession(store AuthStorer, c *gin.Context) error {
-	session := sessions.Default(c)
-	sessionID, ok := session.Get(SessionIDKey).(string)
-	if !ok {
-		return auth.ErrorAuthFailed
+// AuthenticateExternalInitiator authenticates an external initiator request.
+//
+// Implements authMethod
+func AuthenticateExternalInitiator(c *gin.Context, store Authenticator) error {
+	eia := &auth.Token{
+		AccessKey: c.GetHeader(static.ExternalInitiatorAccessKeyHeader),
+		Secret:    c.GetHeader(static.ExternalInitiatorSecretHeader),
 	}
 
-	user, err := store.AuthorizedUserWithSession(sessionID)
+	ei, err := store.FindExternalInitiator(eia)
+	if err != nil {
+		if errors.Is(err, sql.ErrNoRows) {
+			return auth.ErrorAuthFailed
+		}
+
+		return errors.Wrap(err, "finding external intiator")
+	}
+
+	ok, err := bridges.AuthenticateExternalInitiator(eia, ei)
 	if err != nil {
 		return err
 	}
-	c.Set(SessionUserKey, &user)
+	if !ok {
+		return auth.ErrorAuthFailed
+	}
+
+	c.Set(SessionExternalInitiatorKey, ei)
+
 	return nil
 }
 
-var _ authType = AuthenticateBySession
+var _ authMethod = AuthenticateExternalInitiator
 
-func RequireAuth(store AuthStorer, methods ...authType) gin.HandlerFunc {
+// Authenticate is middleware which authenticates the request by attempting to
+// authenticate using all the provided methods.
+func Authenticate(store Authenticator, methods ...authMethod) gin.HandlerFunc {
 	return func(c *gin.Context) {
 		var err error
 		for _, method := range methods {
-			err = method(store, c)
+			err = method(c, store)
 			if err != auth.ErrorAuthFailed {
 				break
 			}
 		}
 		if err != nil {
 			c.Abort()
 			jsonAPIError(c, http.StatusUnauthorized, err)
-		} else {
-			c.Next()
+
+			return
 		}
+
+		c.Next()
+	}
+}
+
+// GetAuthenticatedUser extracts the authentication user from the contextx.
+func GetAuthenticatedUser(c *gin.Context) (*clsessions.User, bool) {
+	obj, ok := c.Get(SessionUserKey)
+	if !ok {
+		return nil, false
 	}
+
+	return obj.(*clsessions.User), ok
```

Comment:
I think this type assertion should be handled because it may panic here. `usr, ok := obj.(*clsessions.User)`